src/engine/sparqlExpressions/ConditionalExpressions.cpp

```diff
@@ -0,0 +1,110 @@
+// Authors: Johannes Kalmbach <kalmbacj@cs.uni-freiburg.de>
+
+#include "NaryExpression.h"
+#include "NaryExpressionImpl.h"
+
+namespace sparqlExpression {
+namespace detail::conditional_expressions {
+using namespace sparqlExpression::detail;
+[[maybe_unused]] auto ifImpl = [](EffectiveBooleanValueGetter::Result condition,
+                                  auto&& i, auto&& e) -> IdOrString {
+  if (condition == EffectiveBooleanValueGetter::Result::True) {
+    return AD_FWD(i);
+  } else {
+    return AD_FWD(e);
+  }
+};
+NARY_EXPRESSION(IfExpression, 3,
+                FV<decltype(ifImpl), EffectiveBooleanValueGetter,
+                   ActualValueGetter, ActualValueGetter>);
+}  // namespace detail::conditional_expressions
+
+// ____________________________________________________
+class CoalesceExpression : public SparqlExpression {
+  std::vector<SparqlExpression::Ptr> children_;
+
+ public:
+  explicit CoalesceExpression(std::vector<SparqlExpression::Ptr> children)
+      : children_{std::move(children)} {};
+  ExpressionResult evaluate(EvaluationContext* context) const override {
+    std::vector<ExpressionResult> childResults;
+    std::ranges::for_each(children_, [&](const auto& child) {
+      childResults.push_back(child->evaluate(context));
+    });
+    std::vector<uint64_t> unboundResults;
+    std::vector<uint64_t> nextUnboundResults;
+    unboundResults.reserve(context->size());
+    nextUnboundResults.reserve(context->size());
+    for (size_t i = 0; i < context->size(); ++i) {
+      unboundResults.push_back(i);
+    }
+    VectorWithMemoryLimit<IdOrString> results{context->_allocator};
+    std::fill_n(std::back_inserter(results), context->size(),
+                IdOrString{Id::makeUndefined()});
+
+    auto visitExpressionResult = [&](SingleExpressionResult auto&& res) {
+      auto gen =
+          detail::makeGenerator(std::move(res), context->size(), context);
+      size_t i = 0;
+      auto nextUnbound = unboundResults.begin();
+      if (nextUnbound == unboundResults.end()) {
+        return;
+      }
+
+      for (auto& el : gen) {
+        if (i == *nextUnbound) {
+          IdOrString val{std::move(el)};
+          if (std::holds_alternative<Id>(val) &&
+              std::get<Id>(val) == Id::makeUndefined()) {
+            nextUnboundResults.push_back(i);
+          } else {
+            results.at(*nextUnbound) = std::move(val);
+          }
+          ++nextUnbound;
+          if (nextUnbound == unboundResults.end()) {
+            return;
+          }
+        }
+        ++i;
+      }
+    };
+
+    std::ranges::for_each(std::move(childResults),
+                          [&](ExpressionResult& child) {
+                            std::visit(visitExpressionResult, std::move(child));
+                            unboundResults = std::move(nextUnboundResults);
+                          });
+    return results;
+  }
+
+  std::string getCacheKey(const VariableToColumnMap& varColMap) const override {
```

Comment:
Do I understand it correctly that the main reason why the implementation of `COALESCE` is so much more involved than that of `IF` is that `COALESCE` has a variable number of arguments?

Reply:
Yes. And as soon as that is the case, we have to choose a column based implementation for efficiency reasons.

Reply:
Once I also implement CONCAT (will look somewhat similar concerning the boilerplate structure) I will reason on whether it is beneficial to factor out some common code for such variadic expressions.